### PR TITLE
Feature/#21 Gson

### LINKED ISSUE #21: JSON Serialisation Library
Should we use Jackson or try a different one?

### PR DESCRIPTION
**Pull request by aanciaes**
Closes #21

### CODE REVIEW

**Comment by monxalo** [comment]
Why did we endup using gson and not kotlinx serialization?

**Comment by aanciaes** [comment]
> Would be nice to have a PR for each library so we can see the overview.
> After we decide which one to use, we can just reject the PRs for the others.

Yes that's a good idea, I'll do that.

File: src/main/kotlin/test/ktortemplate/core/utils/json/PageResponseDeserializer.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import com.google.gson.JsonDeserializer
+import com.google.gson.JsonElement
+import com.google.gson.JsonParseException
+import test.ktortemplate.core.utils.pagination.PageResponse
+import test.ktortemplate.core.utils.pagination.PageResponseLink
+import test.ktortemplate.core.utils.pagination.PageResponseMeta
+import java.lang.reflect.Type
+import java.util.ArrayList
+
+class PageResponseDeserializer<T>(private val clazz: Class<T>) : JsonDeserializer<PageResponse<T>> {
```
Comment (aanciaes) [inline src/main/kotlin/test/ktortemplate/core/utils/json/PageResponseDeserializer.kt:13]: Gson does not handle well the deserialisation of generic types such as PageResponse<T>, so a custom deserialiser must be written.

However, this does not affect serialisation which means that serialised classes are ok, and the jsons returned by the endpoint are the same.
Although, for the tests, the custom deserialiser is necessary to convert from : `PageResponse(meta=PageResponseMeta(page=0, size=10, totalElements=5), data=[{id=1.0, brand=brand10, model=model91}, {id=2.0, brand=brand72, model=model12}…)`
to `PageResponse(meta=PageResponseMeta(page=0, size=10, totalElements=5), data=[Car(id=1, brand=brand10, model=model91), Car(id=2, brand=brand72, model=model12)…)`

File: src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import com.google.gson.Gson
+import com.google.gson.GsonBuilder
+import com.google.gson.reflect.TypeToken
+import test.ktortemplate.core.model.Car
+import test.ktortemplate.core.utils.pagination.PageResponse
+import java.lang.reflect.Type
+
+object JsonSettings {
+
+    val PageResponseCarType: Type = object : TypeToken<PageResponse<Car>>() {}.type
```
Comment (aanciaes) [inline src/main/kotlin/test/ktortemplate/core/utils/json/JsonSettings.kt:12]: The typed PageResponse \<Car\> must be added here, and registered below on the gson mapper

File: src/test/kotlin/test/ktortemplate/core/httphandler/TestRoutes.kt [hunk 3]
```diff
@@ -110,15 +111,16 @@ class TestRoutes : KoinTest {
         // Get all
         with(handleRequest(HttpMethod.Get, "/cars")) {
             response.status() `should be equal to` HttpStatusCode.OK
-            val res: PageResponse<Car> = JsonSettings.mapper.readValue(response.content!!)
+            val res: PageResponse<Car> = JsonSettings.mapper.fromJson(response.content, PageResponseCarType)
```
Comment (aanciaes) [inline src/test/kotlin/test/ktortemplate/core/httphandler/TestRoutes.kt:114]: And used right here when deserialising

**Comment by felix19350** [comment]
@aanciaes performance-wise from what I was able to gather gson is actually slower than jackson. In terms of code this does not seem terribly sexy, is there a case for changing from jackson?

Performance:
https://www.ericdecanini.com/2020/10/13/benchmarking-gson-vs-jackson-vs-moshi-2020/

**Comment by monxalo** [comment]
Jackson is more performant if we use annotation, but is so ugly, i mean:
```kotlin
    @JsonIgnoreProperties(value = ["someid"])
    @JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
    data class PostJackson(
        @JsonProperty("userId")
        val userId: Int,
        @JsonProperty("id") val id: Int,
        @JsonProperty("title") val title: String,
        @JsonProperty("body") val body: String,
        @JsonIgnore val ignoredField: String = ""
    )
``` 

Both Jackson and Gson rely on reflection which will always result in a runtime performance impact.
I'm mostly interested in Moshi and Kotlinx Serialization, since these don't use any reflection and are much more kotlin friendly.

File: src/main/kotlin/test/ktortemplate/core/utils/json/PageResponseDeserializer.kt (hunk 1 above)
Comment (monxalo) [reply src/main/kotlin/test/ktortemplate/core/utils/json/PageResponseDeserializer.kt:13]: I don't this this deserializer is needed, this sample works fine:

```kotlin
val sample = """
            {
	"meta": {"page":0, "size":0, "totalElements":0},
	"links": {"self":"", "first":"", "last": ""},
	"links": {"self":"", "first":"", "last": ""},
	"data": [{
		"id": 1.0,
		"brand": "brand10",
		"model": "model91"
	}, {
		"id": 2.0,
		"brand": "brand72",
		"model": "model12"
	}]
}
        """.trimIndent()

val response = Gson().fromJson<PageResponse<*>>(sample, object : TypeToken<PageResponse<Car>>() {}.type)

println(response)

// Result
PageResponse(meta=PageResponseMeta(page=0, size=0, totalElements=0), data=[Car(id=1, brand=brand10, model=model91), Car(id=2, brand=brand72, model=model12)], links=PageResponseLink(self=, first=, prev=null, next=null, last=))
```

>

**Comment by felix19350** [comment]
For me GSON is a no go. It doesn't bring any advantages. If you agree @aanciaes and @monxalo we can close this and start cleaning up our list of pending PRs

**Comment by monxalo** [comment]
Yes, there are better alternatives, let's close this one.

**Pull request closed without merging**